client/core/bond.go

```diff
@@ -39,58 +40,217 @@ func (c *Core) watchExpiredBonds(ctx context.Context) {
+// tier. NOTE: to minimize bond overlap, an asset.Bonder method could provide
+// this estimate, but it is still very short relative to the entire bond
+// lifetime, which is on the order of months.
+func spendableDelay(net dex.Network) int64 {
+	// We use 3*pendingBuffer as a well-padded estimate of this duration. e.g.
+	// with mainnet, we would use a 90 minute pendingBuffer and a 270 minute
+	// spendableDelay to be robust to periods of slow blocks.
+	return 3 * pendingBuffer(net)
+}
+
+// minBondLifetime gives the minimum bond lifetime (a duration from now until
+// lockTime) that a new bond should use to prevent a tier drop. bondExpiry is in
+// seconds.
+func minBondLifetime(net dex.Network, bondExpiry int64) time.Duration {
+	lifeTimeSec := 2*(pendingBuffer(net)+bondExpiry) + spendableDelay(net) // 2*(p+E)+m
+	return time.Second * time.Duration(lifeTimeSec)
+}
+
+func (c *Core) rotateBonds(ctx context.Context) {
 	// 1. Refund bonds with passed lockTime.
 	// 2. Move bonds that are expired according to DEX bond expiry into
 	//    expiredBonds (lockTime<lockTimeThresh).
-	// 3. TODO: Add bonds to keep N bonds active, according to some optional config.
+	// 3. Add bonds to keep N bonds active, according to target tier and max
+	//    bonded amount, posting before expiry of the bond being replaced.
 	now := time.Now().Unix()
 
 	type bondID struct {
 		assetID uint32
 		coinID  []byte
 	}
 
+	bondKey := c.bondKey()
+	if bondKey != nil {
+		defer bondKey.Zero() // it's a copy
+	}
+
 	for _, dc := range c.dexConnections() {
+		// time.Now().After(time.Unix(bond.LockTime-dc.cfg.BondExpiry))
 		lockTimeThresh := now // in case dex is down, expire (to refund) when lock time is passed
+		var bondExpiry int64
+		bondAmts := make(map[uint32]uint64)
+		bondAssets := make(map[uint32]*msgjson.BondAsset)
 		dc.cfgMtx.RLock()
 		if dc.cfg != nil {
+			bondExpiry = int64(dc.cfg.BondExpiry)
+			for symb, ba := range dc.cfg.BondAssets {
+				id, _ := dex.BipSymbolID(symb)
+				bondAmts[id] = ba.Amt
+				bondAssets[id] = ba
+			}
 			lockTimeThresh += int64(dc.cfg.BondExpiry)
 		}
 		dc.cfgMtx.RUnlock()
+		replaceThresh := lockTimeThresh + pendingBuffer(c.net) // replace before expiry to avoid tier drop
 
-		filterExpiredBonds := func(bond []*db.Bond) (liveBonds []*db.Bond) {
-			for _, bond := range bond {
+		var weak []*db.Bond // may re-post ahead of expiry to maintain tier
+		filterExpiredBonds := func(bonds []*db.Bond) (liveBonds []*db.Bond) {
+			for _, bond := range bonds {
 				if int64(bond.LockTime) <= lockTimeThresh {
 					// This is generally unexpected because auth, reconnect, or a
 					// bondexpired notification should do this first.
 					dc.acct.expiredBonds = append(dc.acct.expiredBonds, bond)
-					c.log.Warnf("Expired bond found: %v", coinIDString(bond.AssetID, bond.CoinID))
+					c.log.Infof("Newly expired bond found: %v (%s)", coinIDString(bond.AssetID, bond.CoinID), unbip(bond.AssetID))
```

Comment:
This directly expired bond is not considered in the calculation of `tierDeficit` below. Partially related to my other comment on `acct.tier` being an int filed rather than a function.

Reply:
Right, I hope this will be come much clearly soon, but you get a `bondexpired` notification shortly after (although possibly before if clocks are set badly), which informs us of the server's understanding of our account tier.  Similarly, the `tierchange` route is used when penalization change the account tier.  And for reconnect handling, the auth/`connect` route also reports tier.  Collectively, we never miss a beat in terms of our effective tier (what server things our tier is).

Reply:
As an example:

```
2023-01-29 19:10:13.582 [INF] CORE: Newly expired bond found: 90287fa1e27ec54c03a94389c7bf22fe63eb62b0e408ef2cbf8078a0c5cc945e:0 (dcr)
2023-01-29 19:10:13.582 [DBG] CORE: Expired bond 90287fa1e27ec54c03a94389c7bf22fe63eb62b0e408ef2cbf8078a0c5cc945e:0 (dcr) refundable in about 3m55s.
2023-01-29 19:10:15.599 [INF] CORE: Received bondexpired notification from 127.0.0.1:17273 for account ad5afeb6189a56487703568485ea26a8db4e11ebbbe1d32dccddcff9bf8b5812...
2023-01-29 19:10:15.600 [INF] CORE: notify: |SUCCESS| (bondpost) BondExpired - New tier = 2 (target = 2).
2023-01-29 19:10:25.797 [DBG] CORE[dcr]: tip change: 13924 (5c764064241eee42fd95e2f450fe129cccd5d5e69902411b09b0fa1e56e3093b) => 13925 (4fcef9f65c942496e24ca8e3e02d21ea10dd417af31ab5406f30043ef2820935)
2023-01-29 19:10:25.797 [TRC] CORE: Processing tip change for dcr
2023-01-29 19:10:25.797 [TRC] CORE: asset 42: total = 4000000000, active = 2000000000
2023-01-29 19:10:33.582 [DBG] CORE: Expired bond 90287fa1e27ec54c03a94389c7bf22fe63eb62b0e408ef2cbf8078a0c5cc945e:0 (dcr) refundable in about 3m35s.
```

Sorry for all the typos in my last commment.

Reply:
Ah. So since the server wouldn't consider the expired bond as part of its `ConnectResult.Tier`, but would still consider the weak bond, we're okay to trust `acct.tier` below in our `tierDeficit`. Do we ever validate that the server is reporting the tier that we expect?

Reply:
> Do we ever validate that the server is reporting the tier that we expect?

For this, I think the missing piece is the violation reporting.  Client software always just does it's best, and generally it's not tracking it's own known violations (e.g. whoops I was offline for a while and I'm pretty sure I messed up on swap X, therefore I expect a -1 tier adjustment).  Not sure we'll ever really do that latter thing, but server's comms should certainly list matches and a specific state at which your account failed.

So to have any hope of functioning, we have to be concerned with how server has classified our account.  A related concept is MaxBondedAmt, because we'll start posting bond as needed (when maintenance is enabled) to keep a target tier.

I suppose ideally we'd have something in client that considers what we believe to be our own conduct state.  That's not unique to bonds though.  With the legacy fee, we just got penalized and client had no judgement on the correctness of that, it was up to the user.  I expect we'll evolve this.